python/tvm/utils/__init__.py

```diff
@@ -0,0 +1,303 @@
+    a = nd.array(np.ones(size, dtype="float32"), device=dev)
+    b = nd.array(np.ones(vec_width * 4 * num_threads(), dtype="float32"), device=dev)
+    times = f.time_evaluator(f.entry_name, dev, repeat=5, number=1)(a, b, num_threads())
+    return size * 4 / times.min  # 4 bytes per float32
+
+
+def roofline_analysis(
+    mod: IRModule, params: Dict[str, nd.NDArray], target: Union[str, Target], dev: Device
+) -> profiling.Report:
+    """
+    Create a profiling report that contains roofline and other estimated
+    statistics from running a module on the VM.
+
+    These statistics are calculated by analyzing the lowered TIR of each
+    operator, so they are estimates of the true values. The statistics are:
+      - Bound: Is the operator memory or compute bound. This is computed by
+        assuming that the operator could perfectly cache all loads -- each byte
+        of memory is only loaded once.
+      - Percent of Theoretical Optimal: What percent of theoretical optimal for
+        the bound. i.e. percent of peak memory bandwidth if memory bound,
+        percent of peak FLOP/s if compute bound.
+      - Unique Loaded Bytes: estimation of the number of byte loaded not
+        counting multiple accesses to the same byte.
+      - Estimated Flops: estimated number of floating point operations.
+      - Arithmetic Intensity: ratio of FLOPs per byte of data.
+      - FLOP/s: floating point operations per second.
+      - Bandwidth: Number of bytes loaded per second.
+
+    Parameters
+    ----------
+    mod : IRModule
+      Uncompiled input module>
+
+    params : Dict[str, nd.NDArray]
+
+    target : Union[str, Target]
+      Target to run on.
+
+    dev : Device
+      Device to run on.
+
+    Returns
+    -------
+
+    report : profiling.Report
+      Profiling report which includes the estimated statistics.
+    """
+    if isinstance(target, str):
+        target = Target(target)
+    peak_bandwidth = estimate_peak_bandwidth(target, dev)
+    peak_flops = estimate_peak_fma_flops(target, dev)
+
+    ridge_point = peak_flops / peak_bandwidth
+
+    all_features = _estimated_features(mod, params, target)
+
+    lib = relay.vm.compile(mod, params=params, target=target)
+    vmexec = profiler_vm.VirtualMachineProfiler(lib, dev)
+
+    args = _create_args(mod, dev)
+    report = vmexec.profile(*args)
+    new_calls = []
+    for call in report.calls:
+        if "Hash" in call.keys():
+            _, features = all_features[call["Hash"]]
+
+            flops = np.sum(features["float_addsub"] + features["float_mul"] + features["float_mad"])
+            unique_loaded_bytes = 0.0
+            # assume no more than 100 buffers
+            for i in range(100):
+                # We could uses loaded bytes, but that accounts for for L1 cache.
+                # If we use unique_bytes, then we are looking at how close we come
+                # to the performance assuming all data is cached perfectly.
```

Comment:
I don't understand why you prefer to make this assumption on data being accessed from the cache perfectly? This will bias you towards fewer loads overall, or a greater arith intensity. Likely I am not understanding this comment.

Reply:
I was thinking of this as a measure of "given the amount of data we must load, how close are we to loading it efficiently." For example, we could be doing many unnecessary passes over the data just in order to get good bandwidth, but a single pass over the data in a suboptimal way may result in faster runtime. It does bias the results a little towards a higher arithmetic intensity, but it is a more accurate measure of how close the program is to optimal bandwidth usage. However, it doesn't match how flops are counted, no maybe it is worth switching. Thought?

Reply:
>"given the amount of data we must load, how close are we to loading it efficiently."

I see okay, so I think you're thinking about this like an upperbound on arithmetic intensity wrt to memory access, ie how far can we push the arithmetic intensity just by efficiently accessing the data. 

I am mostly worried about the scenario in which we don't account for multiple loads of the same bytes in an algorithm. My feeling is that it would be desirable to see the arithmetic intensity increase when the choice of algorithm is changed to reduce the number of loads/stores for the same number of flops. That alone can change the problem from memory bound to compute bound, but it would be deceiving if our algorithmic choice appeared to be compute bound only because the arithmetic intensity wasn't counting repeated loads on the same bytes. 

Said differently, when using unique_bytes the same arithmetic intensity could result for two algorithm choices with the same total flops but different numbers of loads/stores.

Reply:
You've convinced me, I'll switch it over.